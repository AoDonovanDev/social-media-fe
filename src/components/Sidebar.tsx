import * as React from 'react';
import GlobalStyles from '@mui/joy/GlobalStyles';
import Avatar from '@mui/joy/Avatar';
import Box from '@mui/joy/Box';
import Divider from '@mui/joy/Divider';
import IconButton from '@mui/joy/IconButton';
import List from '@mui/joy/List';
import ListItem from '@mui/joy/ListItem';
import ListItemButton, { listItemButtonClasses } from '@mui/joy/ListItemButton';
import ListItemContent from '@mui/joy/ListItemContent';
import Typography from '@mui/joy/Typography';
import Sheet from '@mui/joy/Sheet';
import HomeIcon from '@mui/icons-material/Home';
import ExploreIcon from '@mui/icons-material/Explore';
import LogoutRoundedIcon from '@mui/icons-material/LogoutRounded';
import GroupRoundedIcon from '@mui/icons-material/GroupRounded';
import BrightnessAutoRoundedIcon from '@mui/icons-material/BrightnessAutoRounded';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import AddRoundedIcon from '@mui/icons-material/AddRounded';

import { closeSidebar } from '../util/utils';
import { useNavigate } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from '../app/hooks';
import { logout } from '../features/auth/authSlice';
import ColorSchemeToggle from './ColorSchemeToggle';




function Toggler({
  defaultExpanded = false,
  renderToggle,
  children,
}: {
  defaultExpanded?: boolean;
  children: React.ReactNode;
  renderToggle: (params: {
    open: boolean;
    setOpen: React.Dispatch<React.SetStateAction<boolean>>;
  }) => React.ReactNode;
}) {
  const [open, setOpen] = React.useState(defaultExpanded);
  return (
    <React.Fragment>
      {renderToggle({ open, setOpen })}
      <Box
        sx={[
          {
            display: 'grid',
            transition: '0.2s ease',
            '& > *': {
              overflow: 'hidden',
            },
          },
          open ? { gridTemplateRows: '1fr' } : { gridTemplateRows: '0fr' },
        ]}
      >
        {children}
      </Box>
    </React.Fragment>
  );
}


export default function Sidebar() {

  const authStore = useAppSelector(store => store.auth);
  const dispatch = useAppDispatch();

  const navigate = useNavigate();

  function handleLogOut(e: React.MouseEvent) {
    e.stopPropagation();
    localStorage.removeItem("user");
    dispatch(logout());
    navigate("/login",);
  }

  function handleNavigation(route: string) {
    closeSidebar();
    navigate(route);
  }


  return (
    <Sheet
      className="Sidebar"
      sx={{
        position: { xs: 'fixed', md: 'sticky' },
        transform: {
          xs: 'translateX(calc(100% * (var(--SideNavigation-slideIn, 0) - 1)))',
          md: 'none',
        },
        transition: 'transform 0.4s, width 0.4s',
        zIndex: 10000,
        height: '100dvh',
        width: 'var(--Sidebar-width)',
        top: 0,
        p: 2,
        flexShrink: 0,
        display: 'flex',
        flexDirection: 'column',
        gap: 2,
        borderRight: '1px solid',
        borderColor: 'divider',
      }}
    >
      <GlobalStyles
        styles={(theme) => ({
          ':root': {
            '--Sidebar-width': '220px',
            [theme.breakpoints.up('lg')]: {
              '--Sidebar-width': '240px',
            },
          },
        })}
      />
      <Box
        className="Sidebar-overlay"
        sx={{
          position: 'fixed',
          zIndex: 9998,
          top: 0,
          left: 0,
          width: '100vw',
          height: '100vh',
          opacity: 'var(--SideNavigation-slideIn)',
          backgroundColor: 'var(--joy-palette-background-backdrop)',
          transition: 'opacity 0.4s',
          transform: {
            xs: 'translateX(calc(100% * (var(--SideNavigation-slideIn, 0) - 1) + var(--SideNavigation-slideIn, 0) * var(--Sidebar-width, 0px)))',
            lg: 'translateX(-100%)',
          },
        }}
        onClick={() => closeSidebar()}
      />
      <Box sx={{ display: 'flex', alignItems: 'center', justifyContent: 'space-between' }}>
        <Box sx={{ display: 'flex', gap: 1, alignItems: 'center' }}>
          <IconButton variant="soft" color="primary" size="sm">
            <BrightnessAutoRoundedIcon />
          </IconButton>
          <Typography level="title-lg">Social Media</Typography>
        </Box>
        <ColorSchemeToggle />
      </Box>
      <Divider />
      <Box
        sx={{
          minHeight: 0,
          overflow: 'hidden auto',
          flexGrow: 1,
          display: 'flex',
          flexDirection: 'column',
          [`& .${listItemButtonClasses.root}`]: {
            gap: 1.5,
          },
        }}
      >
        <List
          size="sm"
          sx={{
            gap: 1,
            '--List-nestedInsetStart': '30px',
            '--ListItem-radius': (theme) => theme.vars.radius.sm,
          }}
        >
          <ListItem>
            <ListItemButton
              onClick={
                () => handleNavigation('/feed')
              }
            >
              <HomeIcon />

              <ListItemContent>
                <Typography level="title-sm">Feed</Typography>
              </ListItemContent>
            </ListItemButton>
          </ListItem>

          <ListItem>
            <ListItemButton
              onClick={
                () => handleNavigation('/discover')
              }
            >
              <ExploreIcon />

              <ListItemContent>
                <Typography level="title-sm">Discover</Typography>
              </ListItemContent>
            </ListItemButton>
          </ListItem>


          <ListItem>
            <ListItemButton
              onClick={
                () => handleNavigation('/posts/create')
              }
            >
              <AddRoundedIcon />

              <ListItemContent>
                <Typography level="title-sm">Create Post</Typography>
              </ListItemContent>
            </ListItemButton>
          </ListItem>

          <ListItem nested>
            <Toggler
              defaultExpanded
              renderToggle={({ open, setOpen }) => (
                <ListItemButton onClick={() => setOpen(!open)}>
                  <GroupRoundedIcon />
                  <ListItemContent>
                    <Typography level="title-sm">Users</Typography>
                  </ListItemContent>
                  <KeyboardArrowDownIcon
                    sx={[
                      open ? { transform: 'rotate(180deg)' } : { transform: 'none' },
                    ]}
                  />
                </ListItemButton>
              )}
            >
              <List sx={{ gap: 0.5 }}>
                <ListItem sx={{ mt: 0.5 }}>
                  <ListItemButton
                    onClick={
                      () => handleNavigation('/users/following')
                    }
                  >Following</ListItemButton>
                </ListItem>
                <ListItem>
                  <ListItemButton
                    onClick={
                      () => handleNavigation('/users/followers')
                    }
                  >Followers</ListItemButton>
                </ListItem>
                <ListItem>
                  <ListItemButton

                    onClick={
                      () => handleNavigation('/users/search')
                    }
                  >
                    Search
                  </ListItemButton>
                </ListItem>
              </List>
            </Toggler>
          </ListItem>

        </List>
      </Box>
      <Divider />
      <Box
        sx={{ display: 'flex', gap: 1, alignItems: 'center', cursor: 'pointer' }}
        onClick={() => handleNavigation('/profile')}
      >
        <Avatar
          variant="outlined"
          size="sm"
          src={authStore.auth?.profilePicture as string}
        />
        <Box sx={{ minWidth: 0, flex: 1 }}>
          <Typography level="title-sm" data-testid="fn_ln">{`${authStore.auth?.firstname} ${authStore.auth?.lastname}`}</Typography>
          <Typography level="body-xs">{authStore.auth?.username}</Typography>
        </Box>
        <IconButton size="sm" variant="plain" color="neutral" title="Logout" onClick={(e) => handleLogOut(e)} >
          <LogoutRoundedIcon />
        </IconButton>
      </Box>
    </Sheet>
  );
}

