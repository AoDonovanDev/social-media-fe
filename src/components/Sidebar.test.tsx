import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';
import Sidebar from './Sidebar';
import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../features/auth/authSlice';
import {
  CssVarsProvider,
} from "@mui/joy";


const preloadedState = {
  auth: {
    auth: {
      token: null,
      id: null,
      username: null,
      profilePicture: null,
      bio: null,
      firstname: null,
      lastname:  null,
      exipreDate: undefined
    },
    isLoading: false, 
  },
};

const testStore = configureStore({
  reducer: {
    auth: authReducer,
  },
  preloadedState, 
});

describe('Sidebar Component', () => {
  it('renders without crashing', () => {
    const { container } = render(
      <CssVarsProvider>
      <Provider store={testStore}>
        <Router>
          <Sidebar />
        </Router>
      </Provider>
      </CssVarsProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
