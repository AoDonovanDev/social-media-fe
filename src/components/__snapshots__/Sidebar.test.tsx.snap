// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar Component > renders without crashing 1`] = `
<div>
  <div
    class="MuiSheet-root MuiSheet-variantPlain MuiSheet-colorNeutral Sidebar css-99mksk-JoySheet-root"
  >
    <div
      class="Sidebar-overlay MuiBox-root css-ue4spe"
    />
    <div
      class="MuiBox-root css-1lekzkb"
    >
      <div
        class="MuiBox-root css-1c1kq07"
      >
        <button
          class="MuiIconButton-root MuiIconButton-variantSoft MuiIconButton-colorPrimary MuiIconButton-sizeSm css-dscaku-JoyIconButton-root"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
            data-testid="BrightnessAutoRoundedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10.85 12.65h2.3L12 9zM20 8.69V6c0-1.1-.9-2-2-2h-2.69l-1.9-1.9c-.78-.78-2.05-.78-2.83 0L8.69 4H6c-1.1 0-2 .9-2 2v2.69l-1.9 1.9c-.78.78-.78 2.05 0 2.83l1.9 1.9V18c0 1.1.9 2 2 2h2.69l1.9 1.9c.78.78 2.05.78 2.83 0l1.9-1.9H18c1.1 0 2-.9 2-2v-2.69l1.9-1.9c.78-.78.78-2.05 0-2.83zm-5.91 6.71L13.6 14h-3.2l-.49 1.4c-.13.36-.46.6-.84.6-.62 0-1.05-.61-.84-1.19l2.44-6.86c.2-.57.73-.95 1.33-.95s1.13.38 1.34.94l2.44 6.86c.21.58-.22 1.19-.84 1.19-.39.01-.72-.23-.85-.59"
            />
          </svg>
        </button>
        <p
          class="MuiTypography-root MuiTypography-title-lg css-ar9ysw-JoyTypography-root"
        >
          Social Media
        </p>
      </div>
      <button
        class="MuiIconButton-root MuiIconButton-variantOutlined MuiIconButton-colorNeutral MuiIconButton-sizeSm css-wcii5g-JoyIconButton-root"
        data-screenshot="toggle-mode"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
          data-testid="DarkModeRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.01 3.05C6.51 3.54 3 7.36 3 12c0 4.97 4.03 9 9 9 4.63 0 8.45-3.5 8.95-8 .09-.79-.78-1.42-1.54-.95-.84.54-1.84.85-2.91.85-2.98 0-5.4-2.42-5.4-5.4 0-1.06.31-2.06.84-2.89.45-.67-.04-1.63-.93-1.56"
          />
        </svg>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
          data-testid="LightModeIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 7c-2.76 0-5 2.24-5 5s2.24 5 5 5 5-2.24 5-5-2.24-5-5-5M2 13h2c.55 0 1-.45 1-1s-.45-1-1-1H2c-.55 0-1 .45-1 1s.45 1 1 1m18 0h2c.55 0 1-.45 1-1s-.45-1-1-1h-2c-.55 0-1 .45-1 1s.45 1 1 1M11 2v2c0 .55.45 1 1 1s1-.45 1-1V2c0-.55-.45-1-1-1s-1 .45-1 1m0 18v2c0 .55.45 1 1 1s1-.45 1-1v-2c0-.55-.45-1-1-1s-1 .45-1 1M5.99 4.58c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0s.39-1.03 0-1.41zm12.37 12.37c-.39-.39-1.03-.39-1.41 0-.39.39-.39 1.03 0 1.41l1.06 1.06c.39.39 1.03.39 1.41 0 .39-.39.39-1.03 0-1.41zm1.06-10.96c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0zM7.05 18.36c.39-.39.39-1.03 0-1.41-.39-.39-1.03-.39-1.41 0l-1.06 1.06c-.39.39-.39 1.03 0 1.41s1.03.39 1.41 0z"
          />
        </svg>
      </button>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-horizontal css-1n86vx2-JoyDivider-root"
    />
    <div
      class="MuiBox-root css-an29w6"
    >
      <ul
        class="MuiList-root MuiList-vertical MuiList-variantPlain MuiList-colorNeutral MuiList-sizeSm css-1yzuzoj-JoyList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-colorNeutral MuiListItem-variantPlain css-y25igk-JoyListItem-root"
          data-first-child=""
        >
          <div
            class="MuiListItemButton-root MuiListItemButton-colorNeutral MuiListItemButton-variantPlain css-1xphdof-JoyListItemButton-root"
            data-first-child=""
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
              data-testid="HomeRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1"
              />
            </svg>
            <div
              class="MuiListItemContent-root css-1adno9d-JoyListItemContent-root"
            >
              <p
                class="MuiTypography-root MuiTypography-title-sm css-1q6xxoq-JoyTypography-root"
              >
                Home
              </p>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-colorNeutral MuiListItem-variantPlain css-cl8xxo-JoyListItem-root"
        >
          <div
            class="MuiListItemButton-root MuiListItemButton-colorNeutral MuiListItemButton-variantPlain css-1xphdof-JoyListItemButton-root"
            data-first-child=""
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
              data-testid="AddRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M18 13h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H6c-.55 0-1-.45-1-1s.45-1 1-1h5V6c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1"
              />
            </svg>
            <div
              class="MuiListItemContent-root css-1adno9d-JoyListItemContent-root"
            >
              <p
                class="MuiTypography-root MuiTypography-title-sm css-1q6xxoq-JoyTypography-root"
              >
                Create Post
              </p>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-nested MuiListItem-colorNeutral MuiListItem-variantPlain css-doe3wa-JoyListItem-root"
          data-last-child=""
        >
          <div
            class="MuiListItemButton-root MuiListItemButton-colorNeutral MuiListItemButton-variantPlain css-10xtomd-JoyListItemButton-root"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
              data-testid="GroupRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5s-3 1.34-3 3 1.34 3 3 3m-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5 5 6.34 5 8s1.34 3 3 3m0 2c-2.33 0-7 1.17-7 3.5V18c0 .55.45 1 1 1h12c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5m8 0c-.29 0-.62.02-.97.05.02.01.03.03.04.04 1.14.83 1.93 1.94 1.93 3.41V18c0 .35-.07.69-.18 1H22c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5"
              />
            </svg>
            <div
              class="MuiListItemContent-root css-1adno9d-JoyListItemContent-root"
            >
              <p
                class="MuiTypography-root MuiTypography-title-sm css-1q6xxoq-JoyTypography-root"
              >
                Users
              </p>
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1mixwf2-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
              />
            </svg>
          </div>
          <div
            class="MuiBox-root css-5e5559"
          >
            <ul
              class="MuiList-root MuiList-vertical MuiList-variantPlain MuiList-colorNeutral MuiList-nesting css-1jdlr5h-JoyList-root"
            >
              <li
                class="MuiListItem-root MuiListItem-nesting MuiListItem-colorNeutral MuiListItem-variantPlain css-719v9z-JoyListItem-root"
                data-first-child=""
              >
                <div
                  class="MuiListItemButton-root MuiListItemButton-colorNeutral MuiListItemButton-variantPlain css-1xphdof-JoyListItemButton-root"
                  data-first-child=""
                  role="button"
                  tabindex="0"
                >
                  All
                </div>
              </li>
              <li
                class="MuiListItem-root MuiListItem-nesting MuiListItem-colorNeutral MuiListItem-variantPlain css-cl8xxo-JoyListItem-root"
                data-last-child=""
              >
                <div
                  class="MuiListItemButton-root MuiListItemButton-colorNeutral MuiListItemButton-variantPlain css-1xphdof-JoyListItemButton-root"
                  data-first-child=""
                  role="button"
                  tabindex="0"
                >
                  Following
                </div>
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-horizontal css-1n86vx2-JoyDivider-root"
    />
    <div
      class="MuiBox-root css-sjgw8k"
    >
      <div
        class="MuiAvatar-root MuiAvatar-variantOutlined MuiAvatar-colorNeutral MuiAvatar-sizeSm css-1qb4mhu-JoyAvatar-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-sizeMd MuiAvatar-fallback css-2b7dmi-JoySvgIcon-root-JoyAvatar-fallback"
          data-testid="PersonIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
      <div
        class="MuiBox-root css-1xhmi63"
      >
        <p
          class="MuiTypography-root MuiTypography-title-sm css-1q6xxoq-JoyTypography-root"
        >
          null null
        </p>
        <span
          class="MuiTypography-root MuiTypography-body-xs css-h7h2qx-JoyTypography-root"
        />
      </div>
      <button
        class="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorNeutral MuiIconButton-sizeSm css-1ak5rn2-JoyIconButton-root"
        title="Logout"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-15mrjvf-MuiSvgIcon-root"
          data-testid="LogoutRoundedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M5 5h6c.55 0 1-.45 1-1s-.45-1-1-1H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h6c.55 0 1-.45 1-1s-.45-1-1-1H5z"
          />
          <path
            d="m20.65 11.65-2.79-2.79c-.32-.32-.86-.1-.86.35V11h-7c-.55 0-1 .45-1 1s.45 1 1 1h7v1.79c0 .45.54.67.85.35l2.79-2.79c.2-.19.2-.51.01-.7"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
